riff-raff/app/deployment/actors.scala

```diff
@@ -115,209 +110,310 @@ class DeployMetricsActor extends Actor with Logging {
+
+    case StartDeployment() =>
+      runDeployments(firstDeployments)
+
+    case DeploymentCompleted(deployment) =>
+      log.debug("Deployment completed")
+      markComplete(deployment)
+      nextDeployments(deployment) match {
+        case Deployments(deployments) =>
+          runDeployments(deployments)
+        case FinishPath() =>
+        case FinishDeploy() =>
+          cleanup
+      }
+
+    case DeploymentFailed(deployment, exception) =>
+      log.debug("Deployment failed")
+      markFailed(deployment)
+      if (isExecuting) {
+        log.debug("Failed during deployment but others still running - deferring clean up")
+      } else {
+        cleanup
+      }
+
+    case Terminated(actor) =>
+      failRootContext("DeploymentRunner unexpectedly terminated", new RuntimeException("DeploymentRunner unexpectedly terminated"))
+      log.warn(s"Received terminate from ${actor.path}")
+  }
+
+  private def createContext: DeployContext = {
+    DeployReporter.withFailureHandling(rootReporter) { implicit safeReporter =>
+      import Configuration.artifact.aws._
+      safeReporter.info("Reading deploy.json")
+      val s3Artifact = S3Artifact(record.parameters.build, bucketName)
+      val json = S3Artifact.withZipFallback(s3Artifact) { artifact =>
+        Try(artifact.deployObject.fetchContentAsString()(client).get)
+      }(client, safeReporter)
+      val project = JsonReader.parse(json, s3Artifact)
+      val context = record.parameters.toDeployContext(record.uuid, project, LookupSelector(), safeReporter, client)
+      if (context.tasks.toTaskList.isEmpty)
+        safeReporter.fail("No tasks were found to execute. Ensure the app(s) are in the list supported by this stage/host.")
+      context
+    }
+  }
+
+  private def runDeployments(deployments: List[DeploymentNode]) = {
+    log.debug(s"Running next deployments: $deployments")
+    try {
+      honourStopFlag(rootReporter) {
+        deployments.foreach { deployment =>
+          val deploymentRunner = context.watch(deploymentRunnerFactory(context, s"${record.uuid}-${deployment.pathName}-${deployment.priority}"))
+          deploymentRunner ! DeploymentRunner.RunDeployment(record.uuid, deployment, rootReporter, new DateTime())
+          markExecuting(deployment)
+        }
+      }
+    } catch {
+      case NonFatal(t) => log.error("Couldn't run deployment", t)
+    }
+  }
+
+  private def honourStopFlag(reporter: DeployReporter)(elseBlock: => Unit) {
+    stopFlagAgent().get(record.uuid) match {
+      case Some(userName) =>
+        log.debug("Stop flag set")
+        val stopMessage = s"Deploy has been stopped by $userName"
+        if (!isExecuting) {
+          DeployReporter.failContext(rootReporter, stopMessage, DeployStoppedException(stopMessage))
+          log.debug("Cleaning up")
+          cleanup
+        }
+
+      case None =>
+        elseBlock
+    }
+  }
+
+  @scala.throws[Exception](classOf[Exception])
+  override def postStop(): Unit = {
```

Comment:
Yep - that's fine. I record if a context has already been failed and check in the method. I'll rename the method or make the check explicit.